pallets/parachain-staking/src/lib.rs

```diff
@@ -0,0 +1,1322 @@
+		DelegatorLeft(T::AccountId, BalanceOf<T>),
+		/// Delegator, Amount Locked, Collator, New Total Amount backing
+		/// Collator
+		Delegation(T::AccountId, BalanceOf<T>, T::AccountId, BalanceOf<T>),
+		/// Delegator, Collator, Amount Unstaked, New Total Amount Staked for
+		/// Collator
+		DelegatorLeftCollator(T::AccountId, T::AccountId, BalanceOf<T>, BalanceOf<T>),
+		/// Paid the account (delegator or collator) the balance as liquid
+		/// rewards
+		Rewarded(T::AccountId, BalanceOf<T>),
+		/// Round inflation range set with the provided annual inflation range
+		RoundInflationSet(Perbill, Perbill, Perbill, Perbill),
+		/// Set total selected candidates to this value [old, new]
+		TotalSelectedSet(u32, u32),
+		/// Set blocks per round [current_round, first_block, old, new]
+		BlocksPerRoundSet(RoundIndex, T::BlockNumber, u32, u32),
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_finalize(n: T::BlockNumber) {
+			let mut round = <Round<T>>::get();
+			if round.should_update(n) {
+				// kill snapshot of current round
+				<AtStake<T>>::remove_prefix(round.current);
+				// mutate round
+				round.update(n);
+				// execute all delayed collator exits
+				Self::execute_delayed_collator_exits(round.current);
+				// select top collator candidates for next round
+				let (collator_count, collator_staked, delegator_staked) = Self::select_top_candidates(round.current);
+				// start next round
+				<Round<T>>::put(round);
+
+				Self::deposit_event(Event::NewRound(
+					round.first,
+					round.current,
+					collator_count,
+					collator_staked,
+					delegator_staked,
+				));
+			}
+		}
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn total_selected)]
+	/// The total candidates selected every round
+	type TotalSelected<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn round)]
+	/// Current round index and next round scheduled transition
+	pub type Round<T: Config> = StorageValue<_, RoundInfo<T::BlockNumber>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn delegator_state)]
+	/// Get delegator state associated with an account if account is delegating
+	/// else None
+	type DelegatorState<T: Config> =
+		StorageMap<_, Twox64Concat, T::AccountId, Delegator<T::AccountId, BalanceOf<T>>, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn collator_state)]
+	/// Get collator state associated with an account if account is collating
+	/// else None
+	type CollatorState<T: Config> =
+		StorageMap<_, Twox64Concat, T::AccountId, Collator<T::AccountId, BalanceOf<T>>, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn selected_candidates)]
+	/// The collator candidates selected for the current round
+	type SelectedCandidates<T: Config> = StorageValue<_, Vec<T::AccountId>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn total)]
+	/// Total capital locked by this staking pallet
+	// TODO: Might want to use Struct instead of Tuple
```

Comment:
yes i agree with this. zero overhead cost, but better ergonomics when writing code.